source/common/buffer/buffer_impl.cc

```diff
@@ -12,99 +12,315 @@
+    }
+    return true;
+  }
+
+private:
+  /**
+   * Compute a slice size big enough to hold a specified amount of data.
+   * @param data_size the minimum amount of data the slice must be able to store, in bytes.
+   * @return a recommended slice size, in bytes.
+   */
+  static uint64_t sliceSize(uint64_t data_size) {
+    uint64_t slice_size = 32;
+    while (slice_size < data_size) {
+      slice_size <<= 1;
+      if (slice_size == 0) {
+        // Integer overflow
+        return data_size;
+      }
+    }
+    return slice_size;
+  }
+
+  std::unique_ptr<uint8_t[]> base_;
+
+  /** Offset in bytes from the start of the slice to the start of the Data section */
+  uint64_t data_{0};
+
+  /** Offset in bytes from the start of the slice to the start of the Reserved section */
+  uint64_t reserved_;
+
+  /** Offset in bytes from the start of the slice to the start of the Reservable section */
+  uint64_t reservable_;
+
+  /** Total number of bytes in the slice */
+  uint64_t size_;
+
+  /** Number of outstanding reservations that haven't yet been committed. */
+  unsigned num_reservations_{0};
+};
+
+class UnownedBufferSlice : public BufferSlice {
+public:
+  UnownedBufferSlice(BufferFragment& fragment) : fragment_(fragment), offset_(0) {}
+  ~UnownedBufferSlice() override { fragment_.done(); }
+
+  // BufferSlice
+  const void* data() const override {
+    return static_cast<const uint8_t*>(fragment_.data()) + offset_;
+  }
+
+  void* data() override {
+    return static_cast<uint8_t*>(const_cast<void*>(fragment_.data())) + offset_;
+  }
+
+  void drain(uint64_t size) override {
+    ASSERT(offset_ + size <= fragment_.size());
+    offset_ += size;
+  }
+
+  uint64_t dataSize() const override { return fragment_.size() - offset_; }
+  uint64_t reservableSize() const override { return 0; }
+  Reservation reserve(uint64_t) override { return {nullptr, 0}; }
+  bool commit(const Reservation&, uint64_t) override { return false; }
+
+private:
+  BufferFragment& fragment_;
+
+  /**
+   * Offset in bytes from the start of the fragment to the start of the Data section
+   * (increases with each call to drain()).
+   */
+  uint64_t offset_;
+};
+
+void OwnedImpl::append(const void* data, uint64_t size) {
+  const char* src = static_cast<const char*>(data);
+  while (size != 0) {
+    if (slices_.empty() || slices_.back()->reservableSize() == 0) {
+      slices_.emplace_back(std::make_unique<OwnedBufferSlice>(size));
+    }
+    BufferSlice::Reservation reservation = slices_.back()->reserve(size);
+    ASSERT(reservation.first != nullptr);
+    ASSERT(reservation.second != 0);
+    memcpy(reservation.first, src, reservation.second);
+    slices_.back()->commit(reservation, reservation.second);
+    src += reservation.second;
+    size -= reservation.second;
+  }
 }
 
-void OwnedImpl::add(absl::string_view data) {
-  evbuffer_add(buffer_.get(), data.data(), data.size());
+void OwnedImpl::add(const void* data, uint64_t size) { append(data, size); }
```

Comment:
there's a comment in the base-class saying this method is deprecated; it's worth checking if there are remaining call-sites.

Reply:
I just checked, and there are some remaining call-sites within Envoy itself that depend on this method. `include/envoy/buffer/buffer.h` itself won't even compile with the deprecated method removed.